### PR TITLE
Automatically display trace in databricks

### PR DESCRIPTION
**Pull request by daniellok-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/daniellok-db/mlflow/pull/11524?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11524/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11524
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracing/clients/ipython_wrapper.py
Comment (WeichenXu123) [inline mlflow/tracing/clients/ipython_wrapper.py:40]: So the displayed item might be updated multiple times. Is it possible to only trigger on `display` on the last trace item without invoking `update` ?

I remember the `cell` has an execution completion hook, maybe we can move display to there ?

**Comment by WeichenXu123** [comment]
could you attach a demo notebook ?

**Comment by daniellok-db** [comment]
> could you attach a demo notebook ?

Here, but you won't be able to see the visualization until I merge the frontend PR
https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/3392266048174653

File: mlflow/tracing/clients/ipython_wrapper.py
Comment (daniellok-db) [reply mlflow/tracing/clients/ipython_wrapper.py:40]: i'll investigate!

Comment (daniellok-db) [reply mlflow/tracing/clients/ipython_wrapper.py:40]: i wonder if it's a bit messy to put it in there though—it seems like we need to register a callback, like `print_last_trace()` that will execute on every cell completion. however there might be many cells in which the user is not executing trace-related code, and in order to detect those we might need to check the contents of the cell.

is there any specific concern with having this logic in the `log_trace()` function?

Comment (WeichenXu123) [reply mlflow/tracing/clients/ipython_wrapper.py:40]: OK. Let's keep current PR's code if it doesn't cause performance issue when many updates occurs in one cell

File: mlflow/tracing/clients/ipython_wrapper.py
Comment (daniellok-db) [inline mlflow/tracing/clients/ipython_wrapper.py:7]: @B-Step62 does this look okay? i'm not sure if creating a wrapper is the best way, but I felt it would be convenient later to swap out the client.

Comment (B-Step62) [reply mlflow/tracing/clients/ipython_wrapper.py:7]: @daniellok-db I'm actually wondering if we need to keep the original `InMemoryTraceClient`...? It was particularly created for notebook demo.

Comment (B-Step62) [reply mlflow/tracing/clients/ipython_wrapper.py:7]: It's a bit concerning to extend singleton class... given its mechanism relies on class instance. What about adding the displaying logic in `InMemoryTraceClient` and gate it behind `if IPython.get_ipython() is not None` condition?

Comment (daniellok-db) [reply mlflow/tracing/clients/ipython_wrapper.py:7]: makes sense!

File: mlflow/tracing/clients/local.py
Comment (daniellok-db) [inline mlflow/tracing/clients/local.py:33]: cc @B-Step62 updated to put the logic in the InMemoryClient instead! `_display_trace()` is wrapped entirely in a try catch so it should do nothing in a non-ipython environment

File: mlflow/tracing/clients/local.py
Comment (B-Step62) [inline mlflow/tracing/clients/local.py:59]: (minor, we can do it as follow-up) we can split try catch for importing part and the rest, and show `log.warning()` the actual display errors raised from the latter.

**Review by B-Step62 — APPROVED** [review]
LGTM! Would be great if you test this in E2 as well before merging.

**Comment by daniellok-db** [comment]
> LGTM! Would be great if you test this in E2 as well before merging.

tested and seems to work!

**Pull request merged**
